include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -712,101 +712,148 @@ AffineConstraints<number>::close()
-                    // dofs in our system, we must have a cycle.
-                    ++n_replacements;
-                    Assert(n_replacements / 2 < largest_idx,
-                           ExcMessage("Cycle in constraints detected!"));
+                      // keep track of how many entries we replace in this
+                      // line. If we do more than there are constraints or
+                      // dofs in our system, we must have a cycle.
+                      ++n_replacements;
+                      Assert(n_replacements / 2 < largest_idx,
+                             ExcMessage("Cycle in constraints detected!"));
 #endif
-                  }
-                else
-                  // the DoF that we encountered is not constrained by a
-                  // linear combination of other dofs but is equal to just
-                  // the inhomogeneity (i.e. its chain of entries is
-                  // empty). in that case, we can't just overwrite the
-                  // current entry, but we have to actually eliminate it
-                  {
-                    line.entries.erase(line.entries.begin() + entry);
-                  }
+                    }
+                  else
+                    // the DoF that we encountered is not constrained by a
+                    // linear combination of other dofs but is equal to just
+                    // the inhomogeneity (i.e. its chain of entries is
+                    // empty). in that case, we can't just overwrite the
+                    // current entry, but we have to actually eliminate it
+                    {
+                      entries_to_delete.insert(entry);
+                    }
 
-                line.inhomogeneity += constrained_line.inhomogeneity * weight;
+                  line.inhomogeneity += constrained_line.inhomogeneity * weight;
+                }
 
-                // now that we're here, do not increase index by one but
-                // rather make another pass for the present entry because
-                // we have replaced the present entry by another one, or
-                // because we have deleted it and shifted all following
-                // ones one forward
-              }
-            else
-              // entry not further constrained. just move ahead by one
-              ++entry;
-        }
+            // Now delete the elements we have marked for deletion. Do
+            // so in reverse order so that we compress the array walking
+            // backward from the end without having to keep track that
+            // we have already erased earlier elements (as we would have to
+            // do if we walked the list of entries to delete in forward
+            // order).
+            for (auto it = entries_to_delete.rbegin();
+                 it != entries_to_delete.rend();
+                 ++it)
+              line.entries.erase(line.entries.begin() + *it);
```

Comment:
Does this show up in profiles if there are many entries to delete? If so, would it be better to just iterate through the entries once starting with the first index to delete and move the entries after that accordingly manually?

Reply:
I haven't timed this. The function generally doesn't show up in profiles, so I doubt it. What we're doing here really only applies if a degree of freedom is constrained against another DoF that is itself constrained. This never happens in 2d, for example. In 3d, it can happen on edges if cells with 3 different refinement levels come together there (which I believe we prohibit when we use multigrid), or if three different finite elements come together at an edge. All of these are not common cases that I think are not worth optimizing unless we have concrete evidence that it's worth it.

The patch really only picks of the low-hanging fruit by making sure that we don't keep looking at the same constraints in every iteration because we can determine that some are already finalized.

Reply:
this can also happen in 2D for the corner degrees of freedom with periodic boundary conditions :smiley: